cdap-hbase-compat-1.1/src/main/java/co/cask/cdap/data2/replicationtool/hbase11/LastReplicateTime.java

```
@@ -0,0 +1,68 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.replicationtool.hbase11;
+
+import co.cask.cdap.common.conf.Constants;
+import com.google.protobuf.ByteString;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.hadoop.hbase.CellScanner;
+import org.apache.hadoop.hbase.CoprocessorEnvironment;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.HTableInterface;
+import org.apache.hadoop.hbase.coprocessor.BaseRegionServerObserver;
+import org.apache.hadoop.hbase.coprocessor.ObserverContext;
+import org.apache.hadoop.hbase.coprocessor.RegionServerCoprocessorEnvironment;
+import org.apache.hadoop.hbase.protobuf.generated.AdminProtos;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * HBase Coprocessor that tracks write time of the WAL entries successfully replicated to a Slave Cluster.
+ * For each region the writeTime from the last WAL Entry replicated is updated to the REPLICATION_STATE table.
+ */
+public class LastReplicateTime extends BaseRegionServerObserver {
+  private TableUpdater tableUpdater = null;
+  private static final Log LOG = LogFactory.getLog(LastReplicateTime.class);
+
+  @Override
+  public void start(CoprocessorEnvironment env) throws IOException {
+    LOG.info("LastReplicateTime Start received.");
+    String clusterID = env.getConfiguration().get(Constants.ReplicationStatusTool.CLUSTER_ID);
+    if (clusterID == null) {
+      LOG.error(Constants.ReplicationStatusTool.CLUSTER_ID + " not provided in configuration.");
```

do you want to throw an exception or return after logging error ?

---

agreed. If throwing an exception causes the regionserver to die, then we should return here, and postReplicate should do nothing if tableUpdater is null.

---

I see, you have a null check in postReplicate. Assume you meant to return here then?